### ISSUE TITLE
[RFC] Executable semantics

### ISSUE THREAD

**Issue by Nadrieril** [dhall-lang/dhall-lang#870]
### TL;DR

I propose we write the inference rules of the standard in makam, so that they become executable. That would make experimentation easier and the standard more robust.

### Motivation

Having a reference implementation of the standard has been considered before. The idea would be to have an implementation in a high-level language that is easy to update alongside the standard, and against which we could run our tests.

This would make is easier to experiment with new and/or complex language features without needing to touch one of the actual language implementations, as that requires both knowledge of the language and knowledge of a complex codebase.

This would also ensure that the standard is consistent and does what we expect it to do.

### The proposal

I propose here we do exactly that, in the [makam](https://github.com/astampoulis/makam) metalanguage (see below).
In that language, we can almost directly write the inference rules we currently have, and execute them.

More specifically, we would write the typechecking and normalization semantics in makam, and maintain the current textual inference rules alongside their equivalent in makam.
Both would be similar enough that it shouldn't require a lot more effort to also update the makam rules.

We would then get a way to run our typecheck and normalization tests directly against the spec.

The missing link would be parsing/printing. It's easy to generate a parser from the abnf grammar, but there would remain to write code to convert from the parse tree to an AST, and to pretty-print that AST.

If the proposal gathers enough approval, I'd make a repo to host the makam rule and parsing infrastructure. The eventual goal would be to integrate that repo into `dhall-lang/dhall-lang` and use it to run all tests in CI.

### Makam

[Makam](https://github.com/astampoulis/makam) is a higher-order logic metalanguage, aimed at rapid language prototyping. Here is a good introduction to how it can be used: https://www.tweag.io/posts/2019-11-28-pcf-makam-spec.
As you can see in this link, makam mirrors *very closely* inference rules as would be written in an academic paper, or as currently exist in the dhall standard.

### Drawbacks

This adds a noticeable maintenance burden to the standard. Whereas we previously only edited text files meant for humans, we would now require to additionally edit and test makam rules.

Moreover, we would still need to manually implement and maintain part of the parsing/printing manually.

This also doesn't cover binary encoding/decoding and import resolution, though that could be added later.

### Alternatives

The makam implementation can still be written but never integrate the standard directly. This would lose some of the benefits of having an always up-to-date reference implementation.
I would however personally not be ok for this additional maintenance burden on my own, hence this proposal.

It has been proposed before that we'd write a reference implementation in a more mainstream high-level language like Haskell. The main drawback I see is that updating the standard would become significantly less beginner-approachable and more involved.

It has also been proposed that we'd write a reference implementation in a language that supports [program extraction](https://github.com/coq/coq/wiki/Extraction) like Agda or Coq. Makam is not such a language. This would enable us to generate correct-by-construction implementations.
However in practice, very few languages can be extracted to and it's hard to generate an optimized program from extraction, so I wouldn't expect a lot of implementations to make use of that.

**Comment by Gabriel439** [comment]
@Nadrieril: It's not clear to me why Makam would be significantly more readable to authors of new language bindings.  For example, compare the implementation of `map` in Makam:

```
map Pred [] [].
map Pred (X :: XS) (Y :: YS) :- Pred X Y, map Pred XS YS.
```

... versus `map` in Haskell (written in such a way to resemble the Makam code as closely as possible):

```haskell
map pred [] = []
map pred (x : xs) = y : ys where y = pred x; ys = map pred xs
```

In general, my impression is that fewer people will be able to read a specification written using a logic programming than a functional programming language.

More generally, my concern is that if we use Makam people will avail themselves of features unique to logic programming, which will make translating Dhall to non-logic languages (all of the ones we're interested in) more difficult.  Functional programming style, on the other hand, is essentially portable to any language that has functions (all of the ones we're interested in).

I'm not married to Haskell specifically as the specification language, but I do feel that a functional language would be more appropriate for specification than a logic programming language.  My reasoning is that the main bottleneck to the Dhall ecosystem isn't the difficulty of translating the PLT literature to the Dhall spec, but rather the difficulty of translating the Dhall spec to language bindings.  Writing the spec in a functional programming language optimizes for the latter.

**Comment by Nadrieril** [comment]
Ah, I feel like there's a misunderstanding on what we want to achieve with a reference implementation.

I had been thinking from the assumption that we would keep the existing inference rules as the basis for the semantics of Dhall. A reference implementation would be an extra separate thing that would need to be maintained alongside the inference rules.
In this context, Makam shines: there would be a one-to-one correspondance between inference rules and makam rules. For example:
```
Γ ⊢ f : ∀(x : A₀) → B₀
Γ ⊢ a₀ : A₁
A₀ ≡ A₁
↑(1, x, 0, a₀) = a₁
B₀[x ≔ a₁] = B₁
↑(-1, x, 0, B₁) = B₂
B₂ ⇥ B₃
──────────────────────
Γ ⊢ f a₀ : B₃
```
would be written in makam:
```
typeof (app F A0) B3 :-
    typeof F (arrow x A0 B0),
    typeof A0 A1,
    equiv A0 A1,
    shift 1 x 0 A0 A1,
    subst B0 x A1 B1,
    shift (-1) x 0 B1 B2,
    normalize B2 B3.
```
(we might even be able to generate the textual representation from makam directly :crossed_fingers:).

However I feel like you're proposing something else: to _replace_ the current inference-rule-based semantics with a reference implementation.
In this context I entirely agree with you that having an implementation in a "normal" language would be much better.

So here's the main question I believe: would it be better to have the current inference rules or an implementation in a functional language ?
I'm not entirely sure personally. I'm not even sure which would be more readable. I find inference rules rather legible but I realize I'm not a representative sample.
Personally, the thing I use most when writing my implementation are the tests, since most of what we're doing makes intuitive sense anyways. I only read inference rules when there's a subtler point I'm not sure about. In that context I might prefer inference rules I think.

> I'm not married to Haskell specifically as the specification language, but I do feel that a functional language would be more appropriate for specification than a logic programming language. My reasoning is that the main bottleneck to the Dhall ecosystem isn't the difficulty of translating the PLT literature to the Dhall spec, but rather the difficulty of translating the Dhall spec to language bindings. Writing the spec in a functional programming language optimizes for the latter.

Interesting, I had rather the impression that translating the dhall spec to an implementation wasn't particularly complicated, it's just that it takes a lot of effort. Again, I might not be representative.

**Comment by sjakobi** [comment]
@Nadrieril Could you maybe show us what the typing rules for `merge` would look like in makam? Because that rule is currently rather awkward with its "inventions" of `t₁` and `u₁`. It's also rather difficult to translate to Haskell, so I think it would be nice to have a reference implementation here that's easy to translate for other implementations.

Then there is the imports spec which I believe is the trickiest part the standard. (Is that why so few implementations have implemented it?) So far you don't seem to propose implementing it in makam, possibly because there's no support for HTTP in makam(?). IMHO it's there where a reference implementation could shine most brightly. 

Personally, I would be happy if at least some parts of the standard, like the typing rules for `merge` or the rules for the binary encoding, which I complained about in https://github.com/dhall-lang/dhall-lang/issues/735, would be _replaced_ by working code.

I also believe that if a reference implementation exists separately from a standard document in prose, it will be harder to maintain consistency. In my mind it would be better if we had a single source of truth in a reference implementation that is either well-commented or even written in literate style.

Regarding the language, I think an advantage of Haskell is that the vast majority of the existing standard contributors are already proficient in it. Haskell is also relatively mature, with an established ecosystem, while makam is AFAICT essentially a one-man project.

In sum, I'm much in favor of having a reference implementation, but I'm not yet convinced that makam would be the best language for that purpose.

**Comment by Nadrieril** [comment]
@sjakobi `merge` would give something along the lines of:
```makam
typeof (merge t0 u0) T :-
    typeof t0 (record fields),
    typeof u0 (union variants),
    // Zip variants and fields together. This ensures they have a one-to-one correspondance.
    zip variants fields mergePairs,
    // Ensure all `(variant, field)` pairs are valid and compatible with T
    (p: mergePair -> member p mergePairs -> validMergePair T p).

validMergePair T (arrow A0 T0, presentVariant A1) :-
    equiv A0 A1,
    shift (-1) x 0 T0 T.
validMergePair T (T0, absentVariant) :-
    equiv T T0.
```
This is not quite a direct transcription of the inference rules because handling the record recursively would be cumbersome. Actually, now that I've written this, I think that `validMergePair` would be a useful judgment to define in the standard to make the `merge` rule more legible.

> I think an advantage of Haskell is that the vast majority of the existing standard contributors are already proficient in it.

Hmm, that's a strong point. If most people know Haskell, then a Haskell implementation would be significantly clearer than inference rules.

The more I consider the option of _replacing_ the inference rules, the better it sounds. Then I'd love a literate Haskell standard.

**Comment by sjakobi** [comment]
@Nadrieril Many thanks for the `merge` rule!

I don't quite understand how it ensure that every handler has the same result type. Could you explain?

**Comment by Nadrieril** [comment]
@sjakobi I've rewritten the rule somewhat to have clearer variable names and more comments. As you can see, the `outputs_are_compatible` rule checks that the outputs of each branch of the merge are equivalent to `output_type`.

While writing this makam code, it appears to me that we could write much more legible inference rules if we had access to more abstraction. For example, iterating over the fields of a union/record type. And this is precisely what I could do in makam, but couldn't in the inference rules without introducing some new ad-hoc syntax.
This is achieveing to convince me that inference rules are not an optimal medium for semantics as complex as those we have in Dhall. I'll leave this issue open for a few more days to get more input, but I'm inclined to opening a new one that would propose a literate Haskell standard instead.

**Comment by hanshoglund** [comment]
See also #469

**Comment by Nadrieril** [comment]
I've been convinced against this idea, so I'm closing this now
